src/darwin/Framework/CHIPTests/MTRPerControllerStorageTests.m

```diff
@@ -2172,10 +2172,944 @@ - (void)doTestSubscriptionPoolWithSize:(NSInteger)subscriptionPoolSize
+                    XCTAssertTrue([device _persistedClusterContains:path]);
+                    XCTAssertNotNil([device _getPersistedClusterDataForPath:path]);
+                }
+            }
+        });
+        [subscriptionExpectation fulfill];
+    };
+
+    [device setDelegate:delegate queue:queue];
+
+    [self waitForExpectations:@[ subscriptionExpectation ] timeout:60];
+
+    // Inject a fake attribute report deleting endpoint 2 from the parts list at the root endpoint.
+    dataVersionForPartsList = [NSNumber numberWithUnsignedLongLong:(dataVersionForPartsList.unsignedLongLongValue + 1)];
+    // TODO: try to delete entries from the actual data received in the initila attribute report to make the neunsignedIntegerArrayValue
+    NSArray<NSDictionary<NSString *, id> *> * unsignedIntegerArrayValue = @[
+        @{
+            MTRDataKey : @ {
+                MTRTypeKey : MTRUnsignedIntegerValueType,
+                MTRValueKey : notToBeDeletedEndpoint,
+            }
+        },
+    ];
+
+    NSArray<NSDictionary<NSString *, id> *> * attributeReport = @[ @{
+        MTRAttributePathKey : [MTRAttributePath attributePathWithEndpointID:rootEndpoint clusterID:@(MTRClusterIDTypeDescriptorID) attributeID:@(MTRAttributeIDTypeClusterDescriptorAttributePartsListID)],
+        MTRDataKey : @ {
+            MTRDataVersionKey : dataVersionForPartsList,
+            MTRTypeKey : MTRArrayValueType,
+            MTRValueKey : unsignedIntegerArrayValue,
+        }
+    } ];
+
+    XCTestExpectation * attributeDataReceivedExpectation = [self expectationWithDescription:@"Injected Attribute data received"];
+    XCTestExpectation * reportEndExpectation = [self expectationWithDescription:@"Injected Attribute data report ended"];
+    delegate.onAttributeDataReceived = ^(NSArray<NSDictionary<NSString *, id> *> * attributeReport) {
+        attributeReportsReceived += attributeReport.count;
+        XCTAssert(attributeReportsReceived > 0);
+        [attributeDataReceivedExpectation fulfill];
+    };
+
+    delegate.onReportEnd = ^{
+        // Verify that the data store is still populated with cluster index and cluster data for endpoints 0 and 1 but not for endpoint 2.
+        // Also _persistedClusters and _persistedClusterData has the cluster paths for all clusters in the cluster index for endpoints 0 and 1 but not endpoint 2.
+        NSMutableArray * newTestEndpoints = [testEndpoints mutableCopy];
+        [newTestEndpoints removeObject:toBeDeletedEndpoint];
+        dispatch_sync(self->_storageQueue, ^{
+            XCTAssertTrue([[controller.controllerDataStore _fetchEndpointIndexForNodeID:deviceID] isEqualToArray:newTestEndpoints]);
+            for (NSNumber * endpoint in testEndpoints) {
+                XCTAssertNotNil(initialClusterIndex);
+                for (NSNumber * cluster in [initialClusterIndex objectAtIndex:endpoint.unsignedShortValue]) {
```

Comment:
See comments on previous test about the data structures here.

Reply:
fixed.